File: genium/src/test/java/com/here/genium/validator/visibility/FieldVisibilityValidatorPredicateTest.java
/*
 * Copyright (c) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.ivi.api.validator.visibility;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

import com.here.ivi.api.common.FrancaTypeHelper;
import com.here.ivi.api.model.franca.FrancaDeploymentModel;
import org.franca.core.franca.FField;
import org.franca.core.franca.FType;
import org.franca.core.franca.FTypeRef;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

@RunWith(PowerMockRunner.class)
@PrepareForTest(FrancaTypeHelper.class)
public final class FieldVisibilityValidatorPredicateTest {

  @Mock private FTypeRef francaTypeRef;
  @Mock private FType francaType;
  @Mock private FField francaField;

  @Mock private FrancaDeploymentModel deploymentModel;

  private final FieldVisibilityValidatorPredicate validatorPredicate =
      new FieldVisibilityValidatorPredicate();

  @Before
  public void setUp() {
    MockitoAnnotations.initMocks(this);
    PowerMockito.mockStatic(FrancaTypeHelper.class);

    when(FrancaTypeHelper.getFullName(any())).thenReturn("");

    when(francaField.getType()).thenReturn(francaTypeRef);
    when(francaTypeRef.getDerived()).thenReturn(francaType);
  }

  @Test
  public void validatePublicFieldWithPublicType() {
    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validatePublicFieldWithInternalType() {
    when(deploymentModel.isInternal(francaType)).thenReturn(true);

    assertNotNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateInternalFieldWithPublicType() {
    when(deploymentModel.isInternal(francaField)).thenReturn(true);

    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateInternalFieldWithInternalTypeSamePackage() {
    when(deploymentModel.isInternal(any())).thenReturn(true);
    when(FrancaTypeHelper.haveSamePackage(any(), any())).thenReturn(true);

    assertNull(validatorPredicate.validate(deploymentModel, francaField));
  }

  @Test
  public void validateInternalFieldWithInternalTypeForeignPackage() {
    when(deploymentModel.isInternal(any())).thenReturn(true);
    when(FrancaTypeHelper.haveSamePackage(any(), any())).thenReturn(false);

    assertNotNull(validatorPredicate.validate(deploymentModel, francaField));
  }
}

File: franca-loader/src/test/java/com/here/genium/validator/visibility/FieldVisibilityValidatorPredicateTest.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.validator.visibility

import com.here.genium.franca.FrancaDeploymentModel
import org.eclipse.emf.ecore.EObject
import org.franca.core.franca.FField
import org.franca.core.franca.FModel
import org.franca.core.franca.FType
import org.franca.core.franca.FTypeCollection
import org.franca.core.franca.FTypeRef
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.MockitoAnnotations

@RunWith(JUnit4::class)
class FieldVisibilityValidatorPredicateTest {
    @Mock
    private lateinit var francaTypeRef: FTypeRef
    @Mock
    private lateinit var francaType: FType
    @Mock
    private lateinit var francaField: FField
    @Mock
    private lateinit var francaTypeCollection: FTypeCollection
    @Mock
    private lateinit var francaModel: FModel

    @Mock
    private lateinit var deploymentModel: FrancaDeploymentModel

    private val validatorPredicate = FieldVisibilityValidatorPredicate()

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        `when`(francaModel.name).thenReturn("")
        `when`(francaTypeCollection.name).thenReturn("")
        `when`(francaType.name).thenReturn("")
        `when`(francaField.name).thenReturn("")

        `when`(francaTypeCollection.eContainer()).thenReturn(francaModel)
        `when`(francaType.eContainer()).thenReturn(francaTypeCollection)
        `when`(francaField.eContainer()).thenReturn(francaTypeCollection)

        `when`(francaField.type).thenReturn(francaTypeRef)
        `when`(francaTypeRef.derived).thenReturn(francaType)
    }

    @Test
    fun validatePublicFieldWithPublicType() =
        assertNull(validatorPredicate.validate(deploymentModel, francaField))

    @Test
    fun validatePublicFieldWithInternalType() {
        `when`(deploymentModel.isInternal(francaType)).thenReturn(true)

        assertNotNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateInternalFieldWithPublicType() {
        `when`(deploymentModel.isInternal(francaField)).thenReturn(true)

        assertNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateInternalFieldWithInternalTypeSamePackage() {
        `when`(deploymentModel.isInternal(any<EObject>())).thenReturn(true)

        assertNull(validatorPredicate.validate(deploymentModel, francaField))
    }

    @Test
    fun validateInternalFieldWithInternalTypeForeignPackage() {
        val otherModel = mock(FModel::class.java)
        val otherTypeCollection = mock(FTypeCollection::class.java)
        `when`(otherModel.name).thenReturn("foo")
        `when`(otherTypeCollection.name).thenReturn("")
        `when`(otherTypeCollection.eContainer()).thenReturn(otherModel)
        `when`(francaType.eContainer()).thenReturn(otherTypeCollection)
        `when`(deploymentModel.isInternal(any<EObject>())).thenReturn(true)

        assertNotNull(validatorPredicate.validate(deploymentModel, francaField))
    }
}
